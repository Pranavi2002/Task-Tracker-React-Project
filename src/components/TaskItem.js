import React, { useState } from "react";
import { Card, CardContent, Typography, IconButton, TextField, Stack } from "@mui/material";
import { Edit, Delete, Save, Cancel } from "@mui/icons-material";
import { motion } from "framer-motion";

function TaskItem({ task, onToggle, onDelete, onEdit }) {
  const [isEditing, setIsEditing] = useState(false);
  const [newText, setNewText] = useState(task.text);
  const [editDate, setEditDate] = useState(task.dueDate || "");
  const [errors, setErrors] = useState({});

  const handleSave = () => {
    const todayStr = new Date().toISOString().split("T")[0];
    const validationErrors = {};
    if (!newText.trim()) validationErrors.text = "Task name is required";
    if (editDate && editDate < todayStr) validationErrors.dueDate = "Due date cannot be in the past";

    if (Object.keys(validationErrors).length > 0) {
      setErrors(validationErrors);
      return;
    }

    onEdit(task.id, { text: newText, dueDate: editDate || "" });
    setIsEditing(false);
    setErrors({});
  };

  const isOverdue = !task.completed && task.dueDate && task.dueDate < new Date().toISOString().split("T")[0];

  return (
    <motion.div
      layout
      initial={{ opacity: 0, x: -50 }}
      animate={{ opacity: 1, x: 0 }}
      exit={{ opacity: 0, x: 50 }}
    >
      <Card sx={{ my: 1, bgcolor: "background.paper", color: "text.primary" }}>
        <CardContent sx={{ display: "flex", justifyContent: "space-between", alignItems: "center" }}>
          {isEditing ? (
            <Stack direction={{ xs: "column", sm: "row" }} spacing={1} sx={{ flexGrow: 1, alignItems: "center" }}>
              <TextField
                value={newText}
                onChange={(e) => setNewText(e.target.value)}
                error={!!errors.text}
                helperText={errors.text || ""}
                fullWidth
              />
              <TextField
                type="date"
                value={editDate}
                onChange={(e) => setEditDate(e.target.value)}
                InputLabelProps={{ shrink: true }}
                error={!!errors.dueDate}
                helperText={errors.dueDate || ""}
              />
              <IconButton color="primary" onClick={handleSave}><Save /></IconButton>
              <IconButton color="inherit" onClick={() => setIsEditing(false)}><Cancel /></IconButton>
            </Stack>
          ) : (
            <>
              <Typography
                onClick={() => onToggle(task.id)}
                sx={{
                  textDecoration: task.completed ? "line-through" : "none",
                  color: isOverdue ? "error.main" : "inherit",
                  cursor: "pointer",
                  flexGrow: 1,
                }}
              >
                {task.text} {task.dueDate && `(Due: ${task.dueDate}${isOverdue ? " ⚠️" : ""})`}
              </Typography>
              <Stack direction="row" spacing={1}>
                <IconButton color="primary" onClick={() => setIsEditing(true)}><Edit /></IconButton>
                <IconButton color="error" onClick={() => onDelete(task.id)}><Delete /></IconButton>
              </Stack>
            </>
          )}
        </CardContent>
      </Card>
    </motion.div>
  );
}

export default React.memo(TaskItem);