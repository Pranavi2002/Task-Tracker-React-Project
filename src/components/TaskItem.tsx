// src/components/TaskItem.tsx
import React, { useState } from "react";
import useTheme from "../hooks/useTheme";
import { motion } from "framer-motion";
import { Task } from "../hooks/useTasks";

type TaskItemProps = {
  task: Task;
  onToggle: (id: number) => void;
  onDelete: (id: number) => void;
  onEdit: (id: number, updatedFields: Partial<Task>) => void;
};

const TaskItem: React.FC<TaskItemProps> = ({ task, onToggle, onDelete, onEdit }) => {
  const [isEditing, setIsEditing] = useState(false);
  const [newText, setNewText] = useState(task.text);
  const [editDate, setEditDate] = useState(task.dueDate || "");
  const [errors, setErrors] = useState<{ text?: string; dueDate?: string }>({});

  const { theme } = useTheme();

  const handleSave = () => {
    const validationErrors: { text?: string; dueDate?: string } = {};

    if (!newText.trim()) validationErrors.text = "Task name is required";

    if (editDate) {
      const today = new Date();
      today.setHours(0, 0, 0, 0);

      const due = new Date(editDate);
      if (due < today) validationErrors.dueDate = "Due date cannot be in the past";
    }

    if (Object.keys(validationErrors).length > 0) {
      setErrors(validationErrors);
      return;
    }

    onEdit(task.id, { text: newText, dueDate: editDate || "" });
    setIsEditing(false);
    setErrors({});
  };

  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const taskDue = task.dueDate ? new Date(task.dueDate) : null;
  const isOverdue = !task.completed && taskDue !== null && taskDue < today;

  const editButtonStyle = {
    padding: "5px 10px",
    marginLeft: "5px",
    border: "none",
    borderRadius: "5px",
    cursor: "pointer",
    backgroundColor: "#87CEEB",
    color: "#fff",
    transition: "0.3s",
  };

  const deleteButtonStyle = {
    padding: "5px 10px",
    marginLeft: "5px",
    border: "none",
    borderRadius: "5px",
    cursor: "pointer",
    backgroundColor: "#d32f2f",
    color: "#fff",
    transition: "0.3s",
  };

  const variants = {
    hidden: { opacity: 0, x: -50 },
    visible: { opacity: 1, x: 0 },
    exit: { opacity: 0, x: 50 },
  };

  return (
    <motion.li
      initial="hidden"
      animate="visible"
      exit="exit"
      variants={variants}
      layout
      style={{
        marginBottom: "10px",
        padding: "12px",
        borderRadius: "8px",
        backgroundColor: theme === "light" ? "#fff" : "#333",
        boxShadow:
          theme === "light"
            ? "0 1px 4px rgba(0,0,0,0.1)"
            : "0 1px 4px rgba(0,0,0,0.5)",
        display: "flex",
        justifyContent: "space-between",
        alignItems: "center",
      }}
    >
      {isEditing ? (
        <>
          <input
            type="text"
            value={newText}
            onChange={(e) => setNewText(e.target.value)}
            onKeyDown={(e) => e.key === "Enter" && handleSave()}
            style={{ flexGrow: 1, marginRight: "10px", padding: "5px" }}
          />
          {errors.text && (
            <div style={{ color: "red", fontSize: "12px" }}>{errors.text}</div>
          )}
          <input
            type="date"
            value={editDate}
            onChange={(e) => setEditDate(e.target.value)}
            onKeyDown={(e) => e.key === "Enter" && handleSave()}
            style={{ flexGrow: 1, marginRight: "10px", padding: "5px" }}
          />
          {errors.dueDate && (
            <div style={{ color: "red", fontSize: "12px" }}>{errors.dueDate}</div>
          )}
          <motion.button
            onClick={handleSave}
            style={editButtonStyle}
            whileHover={{ scale: 1.1 }}
            whileTap={{ scale: 0.95 }}
          >
            Save
          </motion.button>
          <motion.button
            onClick={() => setIsEditing(false)}
            style={editButtonStyle}
            whileHover={{ scale: 1.1 }}
            whileTap={{ scale: 0.95 }}
          >
            Cancel
          </motion.button>
        </>
      ) : (
        <>
          <span
            onClick={() => onToggle(task.id)}
            style={{
              textDecoration: task.completed ? "line-through" : "none",
              cursor: "pointer",
              color: isOverdue ? "red" : "inherit",
            }}
          >
            {task.text}{" "}
            {task.dueDate && `(Due: ${task.dueDate}${isOverdue ? " ⚠️ Overdue" : ""})`}
          </span>
          <motion.button
            onClick={() => setIsEditing(true)}
            style={editButtonStyle}
            whileHover={{ scale: 1.1 }}
            whileTap={{ scale: 0.95 }}
          >
            Edit
          </motion.button>
          <motion.button
            onClick={() => onDelete(task.id)}
            style={deleteButtonStyle}
            whileHover={{ scale: 1.1 }}
            whileTap={{ scale: 0.95 }}
          >
            ❌ Delete
          </motion.button>
        </>
      )}
    </motion.li>
  );
};

export default React.memo(TaskItem);