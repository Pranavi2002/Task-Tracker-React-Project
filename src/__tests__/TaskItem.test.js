// src/__tests__/TaskItem.test.js
import React from "react";
import { render } from "@testing-library/react";
import TaskItem from "../components/TaskItem";

// Mock the useTheme hook since TaskItem uses it
jest.mock("../hooks/useTheme", () => () => ({ theme: "light" }));

describe("TaskItem Snapshot", () => {
  // Sample task object for testing
  const mockTask = { id: 1, text: "Test Task", completed: false, dueDate: "2025-08-24" };
  
  // Mock functions for task actions
  const onToggle = jest.fn();
  const onDelete = jest.fn();
  const onEdit = jest.fn();

  it("renders correctly", () => {
    // Render component and take a snapshot
    const { asFragment } = render(
      <TaskItem 
        task={mockTask} 
        onToggle={onToggle} 
        onDelete={onDelete} 
        onEdit={onEdit} 
      />
    );
    // Compare current render to saved snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});