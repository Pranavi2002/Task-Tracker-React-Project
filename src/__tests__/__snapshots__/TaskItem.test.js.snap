// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem Snapshot renders correctly 1`] = `
<DocumentFragment>
  <li
    style="margin-bottom: 10px; padding: 12px; border-radius: 8px; background-color: rgb(255, 255, 255); box-shadow: 0 1px 4px rgba(0,0,0,0.1); display: flex; justify-content: space-between; align-items: center; opacity: 0; transform: translateX(-50px);"
  >
    <span
      style="text-decoration: none; cursor: pointer; color: red;"
    >
      Test Task (Due: 2025-08-24 ⚠️ Overdue)
    </span>
    <button
      style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(135, 206, 235); color: rgb(255, 255, 255); transition: 0.3s;"
      tabindex="0"
    >
      Edit
    </button>
    <button
      style="padding: 5px 10px; margin-left: 5px; border-radius: 5px; cursor: pointer; background-color: rgb(211, 47, 47); color: rgb(255, 255, 255); transition: 0.3s;"
      tabindex="0"
    >
      ❌ Delete
    </button>
  </li>
</DocumentFragment>
`;
